File: web/src/types/recommendation.ts
import type { Track } from "./track";

export type RecommendationCooldownMode = "off" | "soft" | "strict";

export type RecommendationRequest = {
  scene_tag_ids?: number[];
  type_tag_ids?: number[];
  feature_tag_ids?: number[];
  raw_text?: string | null;
  cooldown_mode?: RecommendationCooldownMode;
  limit?: number;
  client?: string | null;
};

export type RecommendationResult = {
  rank: number;
  score: number;
  reason: string;
  explanation?: RecommendationExplanation;
  track: Track;
};

export type RecommendationResponse = {
  request_id: string;
  results: RecommendationResult[];
  exclusions_considered?: string[];
};

export type RevivedTrackCandidate = {
  track: Track;
  last_played_at: string | null;
  playback_count: number;
  days_since_last_played: number | null;
  reason: string;
  tag_summary: string[];
};

export type RevivedTracksResponse = {
  generated_at: string;
  long_unplayed_threshold_days: number;
  never_played_included: boolean;
  candidates: RevivedTrackCandidate[];
};

export type RecommendationExplanationTag = {
  id: number;
  name: string;
  group: string;
};

export type RecommendationExplanationPart = {
  label: string;
  score_delta: number | null;
};

export type RecommendationExplanation = {
  matched_tags?: Record<string, RecommendationExplanationTag[]>;
  boosts?: RecommendationExplanationPart[];
  penalties?: RecommendationExplanationPart[];
  feedback_impacts?: RecommendationExplanationPart[];
  avoidance_reasons?: RecommendationExplanationPart[];
};

File: backend/app/schemas/recommendation.py
from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.track import TrackResponse


CooldownMode = Literal["off", "soft", "strict"]


class RecommendationRequest(BaseModel):
    scenario_tag_ids: list[int] = Field(default_factory=list)
    state_tag_ids: list[int] = Field(default_factory=list)
    type_tag_ids: list[int] = Field(default_factory=list)
    attribute_tag_ids: list[int] = Field(default_factory=list)
    exclude_attribute_tag_ids: list[int] = Field(default_factory=list)
    raw_text: str | None = Field(default=None, max_length=1000)
    cooldown_mode: CooldownMode = "soft"
    limit: int = Field(default=3, ge=1, le=10)
    client: str | None = Field(default=None, max_length=50)


class RecommendationExplanationTag(BaseModel):
    id: int
    name: str
    group: str


class RecommendationExplanationPart(BaseModel):
    label: str
    score_delta: float | None = None


class RecommendationExplanation(BaseModel):
    matched_tags: dict[str, list[RecommendationExplanationTag]] = Field(
        default_factory=dict,
    )
    boosts: list[RecommendationExplanationPart] = Field(default_factory=list)
    penalties: list[RecommendationExplanationPart] = Field(default_factory=list)
    feedback_impacts: list[RecommendationExplanationPart] = Field(default_factory=list)
    avoidance_reasons: list[RecommendationExplanationPart] = Field(default_factory=list)


class RecommendationResult(BaseModel):
    rank: int
    score: float
    reason: str
    explanation: RecommendationExplanation = Field(
        default_factory=RecommendationExplanation,
    )
    track: TrackResponse


class RecommendationResponse(BaseModel):
    request_id: str
    results: list[RecommendationResult]
    exclusions_considered: list[str] = Field(default_factory=list)
